### PR TITLE
fix: Booking Error - cannot be booked at this time

### PR DESCRIPTION
**Pull request by hariombalhara**
## What does this PR do?

Fixes last day not bookable sometimes when future limit is enabled.
[Before fix](https://www.loom.com/share/884ed758109e43108bc8d7e9b49a63ec)
[After fix](https://www.loom.com/share/f03d5a102f174051b35d94806eb8f2a5)

[Loom showing that a similar issue existed earlier as well before Future Limit Rolling Window changes](https://www.loom.com/share/d90ef88771764b8f859f17895e0d0131) - The difference is that in production now, we are showing an additional day(5th june) which is not bookable. Earlier we weren't showing it but last day(4th june) was still unbookable.

## Mandatory Tasks (DO NOT REMOVE)
- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected)
- [x] N/A I have added a Docs issue [here](https://github.com/calcom/docs/issues/new) if this PR makes changes that would require a [documentation change](https://docs.cal.com)
- [x] I have added or modified automated tests that prove my fix is effective or that my feature works (PRs might be rejected if logical changes are not properly tested)

## How should this be tested?
Scenario
1. Create an event with Future Limit of 2 days everything else being default
2. Use timezone America/New York for availability
3. Use Browsing timezone of Pago Pago
4. Try to book last day.

### CODE REVIEW

**Comment by hariombalhara** [comment]
* **#15288** <a href="https://app.graphite.dev/github/pr/calcom/cal.com/15288?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @hariombalhara and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: apps/web/test/lib/getSchedule/futureLimit.timezone.test.ts [hunk 1]
```diff
@@ -29,7 +23,7 @@ function getPeriodTypeData({
   periodEndDate?: Date;
 }) {
   if (type === PeriodType.ROLLING) {
-    if (!periodCountCalendarDays || !periodDays) {
+    if (periodCountCalendarDays === undefined || !periodDays) {
```
Comment (hariombalhara) [inline apps/web/test/lib/getSchedule/futureLimit.timezone.test.ts:26]: To allow `false` value.

File: apps/web/test/lib/getSchedule/futureLimit.timezone.test.ts
Comment (hariombalhara) [inline apps/web/test/lib/getSchedule/futureLimit.timezone.test.ts:1]: Simplified the tests by not using  deprecated `getDate` function and instead mock a particular time to be able to test different sceanrios.

File: apps/web/test/utils/bookingScenario/bookingScenario.ts [hunk 2]
```diff
@@ -814,7 +814,6 @@ export const TestData = {
   },
   schedules: {
     IstWorkHours: {
-      id: 1,
```
Comment (hariombalhara) [inline apps/web/test/utils/bookingScenario/bookingScenario.ts:817]: Wasn't used

File: packages/lib/isOutOfBounds.tsx [hunk 3]
```diff
@@ -51,13 +51,25 @@ export function calculatePeriodLimits({
...
   switch (periodType) {
     case PeriodType.ROLLING: {
       const rollingEndDay = periodCountCalendarDays
-        ? currentTime.add(periodDays, "days").endOf("day")
-        : currentTime.businessDaysAdd(periodDays).endOf("day");
-      return { rollingEndDay, rangeStartDay: null, rangeEndDay: null };
+        ? currentTime.add(periodDays, "days")
+        : currentTime.businessDaysAdd(periodDays);
+      // The future limit talks in terms of days so we take the end of the day here to consider the entire day
+      return { rollingEndDay: rollingEndDay.endOf("day"), rangeStartDay: null, rangeEndDay: null };
```
Comment (hariombalhara) [inline packages/lib/isOutOfBounds.tsx:72]: Doing endOf('day') once.

**Review by hariombalhara — COMMENTED** [review]
Self review done

**Review by joeauyeung — APPROVED** [review]
Awesome find @hariombalhara.

Tested with the organizer set as New York and the booker as Pago Pago. Working as expected. 

I'm approving but I do have a comment about the optimization of filtering slots. It's not blocking but might be something to consider.

**Comment by hariombalhara** [comment]
Regarding this https://github.com/calcom/cal.com/pull/15288#discussion_r1624498200

We were already doing that outOfBound check per timeslot before the future limit improvement. I had tried to optimize that in my previous PR and that approach had this bug that I fixed now.

But it is still better than earlier implementation as we stop as soon as we find the first violation of timeslot. Usually the future limit is a small number of days so we should stop early on.

Starting backwards won't work logically as we have to test for first few bookable time slots, there would always be bookable timeslots at the end but we don't want those. We only want the first ones

**Comment by hariombalhara** [comment]
Thanks for the quick review @joeauyeung 🙏

**Review by CarinaWolli — CHANGES_REQUESTED** [review]
I see a bug with 'Limit future bookings' when testing locally, see loom: https://www.loom.com/share/250ab868612547a8bcec4450b964b94f

**Comment by joeauyeung** [comment]
> I see a bug with 'Limit future bookings' when testing locally, see loom: [loom.com/share/250ab868612547a8bcec4450b964b94f](https://www.loom.com/share/250ab868612547a8bcec4450b964b94f)

Good catch @CarinaWolli! Just tested as well and the time fits the slot but the date is being booked outside of the set range.

**Comment by hariombalhara** [comment]
Great catch @CarinaWolli 

[Just tested this and found it to be an existing behaviour before the future limit improvements](https://www.loom.com/share/6b5f42cb33054b6ebf727cd82ea59fbb). I can take it up as a follow up.

**Review by CarinaWolli — APPROVED** [review]
Can confirm the issue I mentioned already existed before. These changes look good and I created an issue for the other bug: https://github.com/calcom/cal.com/issues/15320

**Pull request merged**